### PR TITLE
feat: Implement member bundle issuance

### PR DESCRIPTION
**Pull request by levansuper**

### CODE REVIEW

File: src/awala.ts [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+  retrieveVeraDnssecChain,
+  selfIssueOrganisationCertificate, serialiseMemberIdBundle,
+} from '@relaycorp/veraid';
+import { MemberModelSchema } from './models/Member.model.js';
+import { Kms } from './utilities/kms/Kms.js';
+import env from 'env-var';
+import { OrgModelSchema } from './models/Org.model.js';
+import { derDeserializePublicKey } from './utilities/derDeserialisePublicKey.js';
+
+const awalaEndpoint = env.get('AWALA_MW').asString();
```
Comment (gnarea) [inline src/awala.ts:20]: I think the word URL is important.

File: src/utilities/derDeserialisePublicKey.ts
Comment (gnarea) [inline src/utilities/derDeserialisePublicKey.ts:1]: Note the the patch I sent had this implemented in `src/utilities/webcrypto.ts`. I think we should keep it there instead of creating a new file.

File: src/awala.ts [hunk 2]
```diff
@@ -33,3 +48,117 @@ export async function createMemberBundleRequest(
     didSucceed: true,
   };
 }
+// string will be substituted with the actual error
+export async function generateMemberBundle(
```
Comment (gnarea) [inline src/awala.ts:52]: I don't think this function belongs in this file: It does, indeed, use the Awala middleware, but just like a [_backing service_](https://12factor.net/backing-services) -- much like MongoDB and the Knative broker (emitter).

The functionality it implements is agnostic of the `POST /awala` endpoint. It's more closely related to the _org member_, so I think we should create a new file like `src/memberBundle.ts` -- analogous to all the other business logic files in `src/*.ts`.

File: src/awala.ts [hunk 3]
```diff
@@ -33,3 +48,117 @@ export async function createMemberBundleRequest(
...
+
+  const memberCertificate = await issueMemberCertificate(
+    member.name,
+    memberCryptoPublicKey,
+    orgCertificate,
+    orgPrivateKey,
+    expiryDate,
+  );
+
+  const dnssecChain = await retrieveVeraDnssecChain(member.orgName, undefined);
```
Comment (gnarea) [inline src/awala.ts:125]: Two things:

- This is quite likely to throw an error, as it connects to a DNS resolver on the Internet, so it'd be good to catch any errors and do a `log.error(...)`. Otherwise, it'd be an uncaught error that would crash the process.
- I _think_ the `undefined` is unnecessary.

**Review by gnarea — CHANGES_REQUESTED** [review]
Overall approach looks good to me.

File: src/utilities/kms/Kms.ts [hunk 4]
```diff
@@ -4,6 +4,16 @@ import { bufferToArrayBuffer } from '../buffer.js';
...
+import { Crypto } from '@peculiar/webcrypto';
+import { CryptoEngine, setEngine } from 'pkijs';
+
+const crypto = new Crypto();
+const cryptoEngine = new CryptoEngine({
+  crypto,
+  name: 'nodeEngine',
+  subtle: crypto.subtle,
+});
+setEngine('nodeEngine', cryptoEngine);
```
Comment (levansuper) [inline src/utilities/kms/Kms.ts:16]: For tests, this is not needed anymore. Do we need this in prod?

File: src/memberBundle.spec.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      expect(issueMemberCertificateParameters[2]).toBe(selfIssueOrganisationCertificateResult);
+      expect(issueMemberCertificateParameters[3]).toBe(calledKeyPair.privateKey);
+      expect(issueMemberCertificateParameters[4].getTime()).toBe(selfIssueOrganisationCertificateParameters[2].getTime());
+
+      expect(retrieveVeraDnssecChain).toHaveBeenCalledOnceWith(ORG_NAME);
+      expect(serialiseMemberIdBundle).toHaveBeenCalledOnceWith(issueMemberCertificateResult, selfIssueOrganisationCertificateResult, retrieveVeraDnssecChainResult);
+      expect(result.result).toBe(serialiseMemberIdBundleResult);
+
+
+      });
```
Comment (levansuper) [inline src/memberBundle.spec.ts:137]: The test looks ugly but this way every method is tested. Maybe we can split the test into multiple to make it more readable

File: src/MemberBundleProblemType.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+export enum MemberBundleProblemType {
+  MEMBER_NOT_FOUND = 'https://veraid.net/problems/member-not-found',
+  MEMBER_NO_NAME = 'https://veraid.net/problems/member-no-name',
+  ORG_NOT_FOUND = 'https://veraid.net/problems/org-not-found',
+  PUBLIC_KEY_NOT_FOUND = 'https://veraid.net/problems/member-public-key-not-found',
+  BUNDLE_GENERATION_ERROR = 'https://veraid.net/problems/member-bundle-generation-error',
+
+}
```
Comment (levansuper) [inline src/MemberBundleProblemType.ts:8]: We could import the errors from the other problem types like:
```
MEMBER_NOT_FOUND = MemberProblemType.MEMBER_NOT_FOUND
```
or we could chenge the method return type:
```
Promise<Result<ArrayBuffer, MemberBundleProblemType>>
```
to
```
Promise<Result<ArrayBuffer, MemberBundleProblemType | MemberProblemType | OrgProblemType>>

```

File: src/testUtils/veraid.mock.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  beforeEach(() => {
+    mockedModule.issueMemberCertificate.mockReset();
+    mockedModule.retrieveVeraDnssecChain.mockReset();
+    mockedModule.selfIssueOrganisationCertificate.mockReset();
+    mockedModule.serialiseMemberIdBundle.mockReset();
+  });
+
+  return mockedModule;
+}
```
Comment (levansuper) [inline src/testUtils/veraid.mock.ts:48]: We can move this file to memberBundle.spec.ts but it will create additional noise

**Review by gnarea — COMMENTED** [review]
Given the state of the code and the fact I don't know what you're planning to improve or not, I only focused on:

- The logic and the security of the functions `createMemberBundleRequest()` and `postToAwala()`.
- The comments you raised.

File: src/awala.ts (hunk 1 above)
Comment (gnarea) [reply src/awala.ts]: I meant we should _add_ the word "URL", not necessarily drop "middleware". Both are key words.

`AWALA_MIDDLEWARE_URL`.

File: src/memberBundle.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    }
+  }
+  const member = await memberModel.findById(memberPublicKey.memberId);
+  if(!member){
+    return {
+      didSucceed: false,
+      reason: MemberBundleProblemType.MEMBER_NOT_FOUND
+    }
+  }
+  if(!member.name){
```
[the remark this answers is not in the document]
Comment (levansuper) [reply src/memberBundle.ts:54]: name is required in issueMemberCertificate so I guess we will need something default

File: src/awala.ts [hunk 9]
```diff
@@ -33,3 +33,35 @@ export async function createMemberBundleRequest(
     didSucceed: true,
   };
 }
+
+export async function postToAwala(
+  data: string,
```
Comment (levansuper) [inline src/awala.ts:38]: What type should data be? Should we allow anything and then stringify or just a string?

File: src/memberBundle.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import { OrgModelSchema } from './models/Org.model.js';
+import { derDeserialisePublicKey } from './utilities/webcrypto.js';
+import { Result } from './utilities/result.js';
+
+export async function generateMemberBundle(
+  publicKeyId: string,
+  options: ServiceOptions,
+): Promise<Result<ArrayBuffer, {
+  shouldRetry: boolean
+}>> {
```
Comment (levansuper) [inline src/memberBundle.ts:23]: I don't really like the response. 
Having Result<ArrayBuffer, boolean> has readability problem when you call the function. So if it fails we get 
```
{
      didSucceed: false,
      reason:true
}
```
And it's really hard to decide what does `reason:true/false` mean

This way it's a bit more visible but still maybe we could create a new type

Comment (gnarea) [reply src/memberBundle.ts:23]: How about rename `FailureResult.reason` to `context`? "Context" is more generic and works in all cases.

Comment (levansuper) [reply src/memberBundle.ts]: That will require to rework almost the whole codebase since it's used a lot

File: src/memberBundle.spec.ts [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+            expect.anything(),
+            expect.anything(),
+          );
+        });
+
+        test('Should be issued with org private key', async () => {
+          await generateMemberBundle(memberPublicKey._id.toString(), serviceOptions);
+
+          // eslint-disable-next-line unicorn/no-unreadable-array-destructuring
+          const [[, , , orgPrivateKey]] = issueMemberCertificate.mock.calls;
```
Comment (levansuper) [inline src/memberBundle.spec.ts:180]: Since this is a test and the rule is not breaking one I think we can disable it

File: src/utilities/kms/Kms.ts (hunk 4 above)
Comment (levansuper) [reply src/utilities/kms/Kms.ts]: during `skaffold dev` error is there without this
```
"err":{"type":"Error","message":"Please call 'setEngine' before call to 'getEngine'","stack":"Error: Please call 'setEngine' before call to 'getEngine'\n    at getEngine (/opt/veraid-authority/node_modules/pkijs/build/index.js:6845:19)\n    at getCrypto (/opt/veraid-authority/node_modules/pkijs/build/index.js:6852:21)\n    at PublicKeyInfo.importKey (/opt/veraid-authority/node_modules/pkijs/build/index.js:3919:41)\n    at Function.issue (/opt/veraid-authority/node_modules/@relaycorp/veraid/src/lib/utils/x509/Certificate.ts:162:42)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at selfIssueOrganisationCertificate (/opt/veraid-authority/node_modules/@relaycorp/veraid/src/lib/pki/organisation.ts:11:23)\n    at generateBundle (/opt/veraid-authority/src/memberBundle.ts:51:26)\n    at generateMemberBundle (/opt/veraid-authority/src/memberBundle.ts:150:24)\n    at aaaa (/opt/veraid-authority/src/memberBundle2.ts:70:9)\n    at triggerBundleRequest (/opt/veraid-authority/src/backgroundQueue/sinks/memberBundleRequestTrigger.sink.ts:72:3)"},"msg":"Please call 'setEngine' before call to 'getEngine'"}
```

**Review by gnarea — CHANGES_REQUESTED** [review]
I reviewed all the code, except for the test files as I'm aware that's work in progress.

Code looks good to me overall. 

I also looked at the new structure of `src/memberBundle.spec.ts`, which looks good to me now.

File: src/memberBundle.spec.ts [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        await generateMemberBundle(memberPublicKey._id.toString(), serviceOptions);
+
+        expect(serialiseMemberIdBundle).toHaveBeenCalledOnceWith(
+          issueMemberCertificateResult,
+          expect.anything(),
+          expect.anything(),
+        );
+      });
+
+      describe('Member bundle serialisation', () => {
```
[the remark this answers is not in the document]
Comment (levansuper) [reply src/memberBundle.spec.ts:217]: Moved the top one but the bottom one tests the `generateMemberBundle` output not the "Member bundle serialisation"

File: src/memberBundle.spec.ts [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+          expect(memberCertificateBuffer).toStrictEqual(memberPublicKeyBuffer);
+        });
+
+        test('Should be issued with org self signed certificate', async () => {
+          await generateMemberBundle(memberPublicKey._id.toString(), serviceOptions);
+
+          expect(issueMemberCertificate).toHaveBeenCalledOnceWith(
+            expect.anything(),
+            expect.anything(),
+            selfIssueCertificateResult,
```
[the remark this answers is not in the document]
Comment (levansuper) [reply src/memberBundle.spec.ts:170]: Did the test and you are right. Will substitute the checks

[the remark this answers is not in the document]
Comment (levansuper) [reply src/memberBundle.spec.ts:170]: But this works so will leave it when we are testing the results
```
        expect(result.result).toStrictEqual(new ArrayBuffer(4));
```

File: src/awala.spec.ts [hunk 14]
```diff
@@ -97,4 +101,45 @@ describe('awala', () => {
...
+      })
+      expect(mockFetch).toHaveBeenNthCalledWith(2, TEST_AWALA_ENDPOINT, {
+        method: 'POST',
+        headers: {
+          [contentTypeHeaderName]: 'application/vnd.veraid.member-bundle',
+          [awalaRecipientHeaderName]: TEST_RECIPIENT_ID
+        },
+        body: awalaPostData,
+      })
+    })
```
Comment (levansuper) [inline src/awala.spec.ts:132]: Since the test is pretty small I made just one test for it. Should we split this into 2 parts? 
1 - should make authorization request
2 - should post data

File: src/memberBundle.spec.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (gnarea) [reply src/memberBundle.spec.ts]: I'm cool with using `ArrayBuffer`s where they do work, but please do not compare uninitialised `ArrayBuffer`s as that's extremely unreliable (not just with Jest): `ArrayBuffer`s are never supposed to be compared or manipulated like that. This is a long/complicated topic that we don't want to discuss here, but if you're curious, check out: https://javascript.info/arraybuffer-binary-arrays

Instead, make sure to actually populate the `ArrayBuffer`s. This is what I've done in other repos:

1. Create/update `src/testUtils/buffer.ts` and add the following:
   ```ts
   function bufferToArrayBuffer(buffer: Uint8Array): ArrayBuffer {
     return buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);
   }

   export function stringToArrayBuffer(string: string): ArrayBuffer {
     return bufferToArrayBuffer(Buffer.from(string));
   }
   ```
2. Initialise every `ArrayBuffer` in the test suite; e.g.:
   ```ts
selfIssueCertificateResult = stringToArrayBuffer('org certificate');
   ```

File: src/awala.spec.ts (hunk 14 above)
Comment (gnarea) [reply src/awala.spec.ts]: Yes, please. I think it should be split. I should be able to understand the requirements from the test suite.

File: package.json [hunk 15]
```diff
@@ -34,7 +35,8 @@
     "is-valid-domain": "^0.1.6",
     "json-schema-to-ts": "^2.8.0",
     "mongoose": "^7.0.4",
-    "pino": "^8.11.0"
+    "pino": "^8.11.0",
+    "pkijs": "^3.0.14"
```
Comment (gnarea) [inline package.json:39]: `pkijs` and all other code related to the workaround should be removed once I've fixed the root issue.

File: src/memberBundle.ts [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+import type { ServiceOptions } from './serviceTypes.js';
+import { MemberPublicKeyModelSchema } from './models/MemberPublicKey.model.js';
+import { MemberModelSchema } from './models/Member.model.js';
+import { Kms } from './utilities/kms/Kms.js';
+import { OrgModelSchema } from './models/Org.model.js';
+import { derDeserialisePublicKey } from './utilities/webcrypto.js';
+import type { Result } from './utilities/result.js';
+
+const CERTIFICATE_EXPIRY_DAYS = 90;
+interface CreateBundleInput {
```
Comment (gnarea) [inline src/memberBundle.ts:20]:
```suggestion
interface BundleCreationInput {
```

Following our code conventions.

File: src/memberBundle.ts [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+  let dnssecChain;
+  try {
+    dnssecChain = await retrieveVeraDnssecChain(orgName);
+  } catch (err) {
+    logger.warn(
+      {
+        memberPublicKeyId,
+        err,
+      },
+      'Failed to retrieve dnssec chain',
```
Comment (gnarea) [inline src/memberBundle.ts:70]:
```suggestion
      'Failed to retrieve DNSSEC chain',
```

File: src/memberBundle.ts [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+    memberName,
+    memberCryptoPublicKey,
+    orgCertificate,
+    orgPrivateKey,
+    expiryDate,
+  );
+
+  let dnssecChain;
+  try {
+    dnssecChain = await retrieveVeraDnssecChain(orgName);
```
Comment (gnarea) [inline src/memberBundle.ts:63]: Sorry, I missed this in the review I did last night:

We should be retrieving the DNSSEC chain before we try to do anything with the KMS: The DNSSEC chain retrieval is expensive CPU- and bandwidth-wise, as well as likely to fail due to the **many** HTTP requests it makes behind the scenes.

So we shouldn't try to use the KMS until we actually have a DNSSEC chain: KMS operations are expensive -- financially and computationally.

Please also write a test for this so we won't introduce a regression in the future.

TL;DR: Move `retrieveVeraDnssecChain()` to the top of this function.

**Review by gnarea — CHANGES_REQUESTED** [review]
Nearly there

File: src/memberBundle.ts (hunk 10 above)
Comment (levansuper) [reply src/memberBundle.ts]: This rework will take place in a dedicated PR. #128

File: src/backgroundQueue/server.spec.ts [hunk 19]
```diff
@@ -51,5 +56,47 @@ describe('makeQueueServer', () => {
...
     });
+
+    describe('makeQueueServerPlugin', () => {
+      const setEnvVars = configureMockEnvVars(REQUIRED_ENV_VARS);
+      const mockFastify: FastifyInstance = {
+        addContentTypeParser: jest.fn(),
+        getDefaultJsonParser: jest.fn(),
+        get: jest.fn(),
+        post: jest.fn(),
+      } as any;
```
Comment (gnarea) [inline src/backgroundQueue/server.spec.ts:67]: Why don't we borrow the approach used elsewhere in this repo? We used all of those methods in other places and never had to mock Fastify to do so.

File: src/backgroundQueue/server.spec.ts [hunk 20]
```diff
@@ -51,5 +56,47 @@ describe('makeQueueServer', () => {
       expect(response.statusCode).toBe(HTTP_STATUS_CODES.BAD_REQUEST);
       expect(response.json()).toHaveProperty('type', QueueProblemType.UNSUPPORTED_EVENT);
     });
+
+    describe('makeQueueServerPlugin', () => {
+      const setEnvVars = configureMockEnvVars(REQUIRED_ENV_VARS);
```
Comment (gnarea) [inline src/backgroundQueue/server.spec.ts:61]: Note that this will create a second mock of the env vars -- each with its own set of `beforeAll()`s and `beforeEach()`s. That may lead to race conditions that may be hard to debug.

We have `getTestServerFixture`. Let's get the mocker from there.

File: src/backgroundQueue/sinks/memberBundleRequest.sink.spec.ts [hunk 21]
```diff
@@ -144,7 +146,7 @@ describe('memberBundleIssuance', () => {
         expect(mockPostToAwala).toHaveBeenCalledOnceWith(
           expect.anything(),
           expect.anything(),
-          new URL(envVars.AWALA_MIDDLEWARE_ENDPOINT!),
+          new URL(REQUIRED_ENV_VARS.AWALA_MIDDLEWARE_ENDPOINT),
```
Comment (gnarea) [inline src/backgroundQueue/sinks/memberBundleRequest.sink.spec.ts:149]: We don't need the whole `REQUIRED_ENV_VARS` and for consistency with other tests we should simply do:

```suggestion
          new URL(AWALA_MIDDLEWARE_ENDPOINT),
```

**Review by gnarea — CHANGES_REQUESTED** [review]

File: src/awala.ts (hunk 2 above)
Comment (levansuper) [reply src/awala.ts]: This was moved to memberBundle.ts - https://github.com/relaycorp/veraid-authority/pull/119/files#diff-5006ad038ccf0df13d5fba8db8001b0fad11d2deb7c5c40fb830dc68e101ab0bR78

File: src/backgroundQueue/sinks/memberBundleRequest.sink.ts [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+  event: CloudEvent<unknown>,
+  options: SinkOptions,
+): Promise<void> {
+  options.logger.debug({ eventId: event.id }, 'Starting member bundle request trigger');
+
+  const validatedData = validateMessage(event.data, MEMBER_BUNDLE_REQUEST_PAYLOAD);
+  if (typeof validatedData === 'string') {
+    options.logger.info(
+      { eventId: event.id, validationError: validatedData },
+      'Malformed event data',
```
[the remark this answers is not in the document]
Comment (levansuper) [reply src/backgroundQueue/sinks/memberBundleRequest.sink.ts:22]: missed this

File: src/backgroundQueue/server.spec.ts (hunk 19 above)
Comment (levansuper) [reply src/backgroundQueue/server.spec.ts:67]: Took an example from here.
https://github.com/relaycorp/veraid-authority/blob/main/src/utilities/fastify/plugins/jwksAuthentication.spec.ts

Comment (gnarea) [reply src/backgroundQueue/server.spec.ts:71]: But we're testing routes here.

Comment (gnarea) [reply src/backgroundQueue/server.spec.ts:71]: Mocking Fastify should only be done in exceptional circumstances (e.g., plugins), not when testing routes.

**Review by gnarea — APPROVED** [review]
Nice one, thanks!
